lib/ansible/modules/cloud/amazon/elasticache_replication_group.py

```diff
@@ -0,0 +1,585 @@
+                waiter = connection.get_waiter('cache_cluster_available')
+                waiter.wait(CacheClusterId=node_group['CacheClusterId'])
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, "Couldn't reboot cache cluster %s" % node_group['CacheClusterId'])
+
+    result = check_for_rep_group(module, connection)
+    module.exit_json(changed=changed, **camel_dict_to_snake_dict(result))
+
+
+def modify_rep_group(module, connection, rep_group):
+
+    changed = False
+
+    params = dict()
+    warnings = list()
+
+    # The output keys of describe_replication_groups are different
+    # to the input keys of modify_replication_group
+    rep_group['ReplicationGroupDescription'] = rep_group['Description']
+
+    rep_group_params = {
+        'description': 'ReplicationGroupDescription',
+        'primary_cluster_id': 'PrimaryClusterId',
+        'node_type': 'CacheNodeType',
+    }
+    cache_cluster_params = {
+        'cache_engine_version': 'EngineVersion',
+        'cache_parameter_group': 'CacheParameterGroupName',
+        'cache_subnet_group_name': 'CacheSubnetGroupName',
+        'cache_security_groups': 'CacheSecurityGroupNames',
+        'preferred_maintenance_window': 'PreferredMaintenanceWindow',
+        'notification_topic_arn': 'NotificationTopicArn',
+        'auto_minor_version_upgrade': 'AutoMinorVersionUpgrade',
+        'snapshot_retention_limit': 'SnapshotRetentionLimit',
+        'snapshot_window': 'SnapshotWindow',
+    }
+
+    cache_cluster_id = rep_group['MemberClusters'][0]
+    try:
+        cache_clusters = connection.describe_cache_clusters(CacheClusterId=cache_cluster_id)['CacheClusters']
+        cache_cluster = cache_clusters[0]
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        if e.response['Error']['Code'] == 'CacheClusterNotFound':
+            cache_cluster = dict()
+        else:
+            module.fail_json_aws(e, msg="Couldn't describe cache cluster")
+
+    for k, v in rep_group_params.items():
+        if module.params.get(k) is not None and rep_group[v] != module.params.get(k):
+            params[v] = module.params.get[k]
+
+    for k, v in cache_cluster_params.items():
+        if module.params.get(k) is not None:
+            if v in cache_cluster:
+                if cache_cluster.get(v) != module.params.get(k):
+                    params[v] = module.params.get(k)
+            else:
+                warnings.append("Cannot determine current cluster value for parameter "
+                                "%s. Blindly setting it" % k)
+                params[v] = module.params.get(k)
+
+    # data structure of returned security groups from describe_cache_clusters is
+    # different to that expected by modify_replication_group
+    if set(module.params['security_group_ids']) != set([sg['SecurityGroupId'] for sg in cache_cluster['SecurityGroups']]):
```

Comment:
In my testing, I've found that if a cluster doesn't have any security groups assigned to it, this key will not exist.  This `if` statement then returns an error complaining about this `NoneType` not being iterable.  I did something like this to get around it.  This probably isn't the best way to do it but it works for me:

```
    # data structure of returned security groups from describe_cache_clusters is
    # different to that expected by modify_replication_group
    if 'SecurityGroups' in cache_cluster:
        if set(module.params['security_group_ids']) != set([sg['SecurityGroupId'] for sg in cache_cluster['SecurityGroups']]):
            params['SecurityGroupIds'] = module.params['security_group_ids']

    if 'SecurityGroups' not in cache_cluster:
        if module.params['security_group_ids']:
            params['SecurityGroupIds'] = module.params['security_group_ids']
```

Reply:
Fixed